subxt/src/utils.rs

```diff
@@ -96,3 +96,137 @@ unsafe impl<T> Sync for PhantomDataSendSync<T> {}
+                    const FORMAT: StoreFormat = StoreFormat::$ty;
+                    const BITS: u32 = <$wrapped>::BITS;
+                }
+            )*
+        };
+    }
+
+    macro_rules! order {
+        ($ident: ident; $($ty: ident),*) => {
+            /// Associates `bitvec::order::BitOrder` trait with corresponding, type-erased `scale_bits::OrderFormat` enum.
+            ///
+            /// Used to decode bit sequences in runtime by providing `scale_bits::OrderFormat` using
+            /// `bitvec`-like type type parameters.
+            pub trait $ident {
+                /// Corresponding `scale_bits::OrderFormat` value.
+                const FORMAT: OrderFormat;
+            }
+
+            $(
+                #[doc = concat!("Type-level value that corresponds to `scale_bits::OrderFormat::", stringify!($ty), "` at run-time")]
+                #[doc = concat!(" and `bitvec::order::BitOrder::", stringify!($ty), "` at the type level.")]
+                #[derive(Clone, Debug, PartialEq, Eq)]
+                pub enum $ty {}
+                impl $ident for $ty {
+                    const FORMAT: OrderFormat = OrderFormat::$ty;
+                }
+            )*
+        };
+    }
+
+    store!(BitStore; (U8, u8), (U16, u16), (U32, u32), (U64, u64));
+    order!(BitOrder; Lsb0, Msb0);
+
+    /// Constructs a run-time format parameters based on the corresponding type-level parameters.
+    pub fn bit_format<Store: BitStore, Order: BitOrder>() -> Format {
+        Format {
+            order: Order::FORMAT,
+            store: Store::FORMAT,
+        }
+    }
+
+    /// `scale_bits::Bits` generic over the bit store (`u8`/`u16`/`u32`/`u64`) and bit order (LSB, MSB) types.
+    #[derive(Debug, Clone, PartialEq, Eq)]
+    pub struct DecodedBits<Store: BitStore, Order: BitOrder>(
+        pub Bits,
+        pub PhantomData<Store>,
+        pub PhantomData<Order>,
+    );
+
+    impl<Store: BitStore, Order: BitOrder> codec::Decode for DecodedBits<Store, Order> {
+        fn decode<I: Input>(input: &mut I) -> Result<Self, codec::Error> {
+            /// Equivalent of `BitSlice::MAX_BITS` on 32bit machine.
+            const ARCH32BIT_BITSLICE_MAX_BITS: usize = 0x1fff_ffff;
+
+            let Compact(bits) = <Compact<u32>>::decode(input)?;
+            // Otherwise it is impossible to store it on 32bit machine.
+            if bits as usize > ARCH32BIT_BITSLICE_MAX_BITS {
+                return Err("Attempt to decode a BitVec with too many bits".into())
+            }
+            // NOTE: Replace with `bits.div_ceil(Store::BITS)` if `int_roundings` is stabilised
+            let bytes_needed =
+                (bits / Store::BITS) as usize + (bits % Store::BITS != 0) as usize;
+
+            // NOTE: We could reduce allocations if it would be possible to directly
+            // decode from an `Input` type using a custom format (rather than default <u8, Lsb0>)
+            // for the `Bits` type.
```

Comment:
It definitely feels from this PR that we could improve `scale_bits::decode_using_format_from` to work with that `Input` trait rather than accepting an array of bytes; it'd be nice to be able to condense all of this decode logic into a few lines much like the `encode` version below, and the only reason that you can't is that the interface in scale-bits isn't quite right :)

Reply:
Opened an issue for this here: https://github.com/paritytech/scale-decode/issues/10